### PR TITLE
Notmnist lowmem

### PR DESCRIPTION
**Pull request by rthouvenin**
As explained in this [Udacity thread](https://discussions.udacity.com/t/deep-learning-notmnist-data-curation/45051) and as requested by Vincent in Issue #936, I am issuing this PR to update the way the notMNIST dataset is loaded into memory, so that it requires less memory.

The idea is to load the dataset in pieces, curate each piece separately, and merge them.

### CODE REVIEW

**Comment by rthouvenin** [comment]
I signed it!

File: tensorflow/examples/udacity/1_notmnist.ipynb [hunk 1]
```diff
@@ -283,83 +283,140 @@
...
+        "    \n",
+        "    print('Full dataset tensor:', dataset.shape)\n",
+        "    print('Mean:', np.mean(dataset))\n",
+        "    print('Standard deviation:', np.std(dataset))\n",
+        "    return dataset\n",
+        "        \n",
+        "def load(data_folders, min_num_images, num_classes):\n",
+        "  dataset_names = []\n",
+        "  for folder in data_folders:\n",
+        "    dataset = load_letter(folder, min_num_images / num_classes)\n",
```
Comment (vincentvanhoucke) [inline tensorflow/examples/udacity/1_notmnist.ipynb]: For clarity, I would slightly prefer load() to be passed a min_num_images_per_class, to be forwarded directly to load_letter()

File: tensorflow/examples/udacity/1_notmnist.ipynb [hunk 2]
```diff
@@ -496,22 +502,61 @@
         "outputId": "8af66da6-902d-4719-bedc-7c9fb7ae7948"
       },
       "source": [
+        "def make_arrays(nb_rows, img_size):\n",
+        "    if nb_rows:\n",
+        "        dataset = np.ndarray((nb_rows, img_size, img_size), dtype=np.float32)\n",
+        "        labels = np.ndarray(nb_rows, dtype=np.int32)\n",
+        "    else:\n",
```
Comment (vincentvanhoucke) [inline tensorflow/examples/udacity/1_notmnist.ipynb]: Why do you require this else clause here? I would prefer simpler.

File: tensorflow/examples/udacity/1_notmnist.ipynb [hunk 3]
```diff
@@ -496,22 +502,61 @@
...
+        "        dataset, labels = None, None\n",
+        "    return dataset, labels\n",
+        "\n",
+        "def merge_datasets(pickle_files, train_size, valid_size=0):\n",
```
Comment (vincentvanhoucke) [inline tensorflow/examples/udacity/1_notmnist.ipynb]: default to zero goes against my pestering that having a validation set is a good thing :) No default is fine.

File: tensorflow/examples/udacity/1_notmnist.ipynb [hunk 4]
```diff
@@ -496,22 +502,61 @@
...
+        "    num_classes = len(pickle_files)\n",
+        "    valid_dataset, valid_labels = make_arrays(valid_size, image_size)\n",
+        "    train_dataset, train_labels = make_arrays(train_size, image_size)\n",
+        "    vsize_per_class = int(valid_size / num_classes)\n",
```
Comment (vincentvanhoucke) [inline tensorflow/examples/udacity/1_notmnist.ipynb]: I think this should be
vsize_per_class = valid_size // num_classes
for Python 3 compatibility.
Same below.

File: tensorflow/examples/udacity/1_notmnist.ipynb [hunk 5]
```diff
@@ -496,22 +502,61 @@
...
-        "print('Validation', valid_dataset.shape, valid_labels.shape)"
+        "print('Training:', train_dataset.shape, train_labels.shape)\n",
+        "print('Validation:', valid_dataset.shape, valid_labels.shape)\n",
+        "print('Testing:', test_dataset.shape, test_labels.shape)"
       ],
       "outputs": [
         {
           "output_type": "stream",
           "text": [
             "Training (200000, 28, 28) (200000,)\n",
```
Comment (vincentvanhoucke) [inline tensorflow/examples/udacity/1_notmnist.ipynb]: Have you verified that subsequent assignments run well and give sensible results with this default training set size?

File: tensorflow/examples/udacity/1_notmnist.ipynb (hunk 3 above)
Comment (rthouvenin) [inline tensorflow/examples/udacity/1_notmnist.ipynb]: This and the test in `make_arrays` is so that the same function can be called for the training data and the test data. I can:

 1. Keep it like this
 2. Remove the default and make it explicit in the call for the test set
 3. Move the size test from `make_array` to `merge_datasets`
 4. Have a different function for the training+validation set and the test set
 5. Something else?

File: tensorflow/examples/udacity/1_notmnist.ipynb (hunk 4 above)
Comment (rthouvenin) [inline tensorflow/examples/udacity/1_notmnist.ipynb]: The call to `int()` make it Python3-compatible, but you're right it is clearer with //.

File: tensorflow/examples/udacity/1_notmnist.ipynb (hunk 5 above)
Comment (rthouvenin) [inline tensorflow/examples/udacity/1_notmnist.ipynb]: I haven't had the time yet to do the subsequent assignments. But this default training size is already in the current notebook.

File: tensorflow/examples/udacity/1_notmnist.ipynb (hunk 3 above)
Comment (vincentvanhoucke) [inline tensorflow/examples/udacity/1_notmnist.ipynb]: You're right, it's simpler this way.

**Comment by vincentvanhoucke** [comment]
Cool, can you wrap the couple of lines that go past the edge and squash all the commits? I'll run a test in my local env this weekend or Monday and pull it. Thanks again!

**Comment by vincentvanhoucke** [comment]
Merged, thanks!
